src/bundle.c

```diff
@@ -278,6 +297,85 @@ static enum swupd_code validate_bundle(const char *bundle, struct manifest *mom,
 	return ret;
 }
 
+static void print_remove_summary(unsigned int requested, unsigned int bad, unsigned int total_removed)
+{
+	int deps_removed;
+
+	if (bad > 0) {
+		print("\nFailed to remove %i of %i bundles\n", bad, requested);
+	} else {
+		print("\nSuccessfully removed %i bundle%s\n", requested, (requested > 1 ? "s" : ""));
+	}
+
+	deps_removed = total_removed + bad - requested;
+	if (deps_removed > 0) {
+		if (cmdline_option_force) {
+			print("%i bundle%s\n", deps_removed, deps_removed > 1 ? "s that depended on the specified bundle(s) were removed" : " that depended on the specified bundle(s) was removed");
+		} else {
+			print("%i bundle%s\n", deps_removed, deps_removed > 1 ? "s that were installed as a dependency were removed" : " that was installed as a dependency was removed");
+		}
+	}
+}
+
+static void get_removable_dependencies(struct manifest *mom, struct list **bundles_to_remove)
+{
+	char *bundle;
+	struct list *iter;
+	struct list *candidates_for_removal = NULL;
+	struct manifest *m;
+
+	/* for each of the bundles requested to be deleted
+	 * get a list of their dependencies we are allowed to remove */
+	for (iter = *bundles_to_remove; iter; iter = iter->next) {
+		m = iter->data;
+		bundle = m->component;
+		recurse_dependencies(mom, bundle, &candidates_for_removal, is_deletable_dependency);
+	}
+
+	/* get a list of dependencies still required by bundles
+	 * that won't be removed */
+	struct list *required_bundles = NULL;
+	for (iter = mom->submanifests; iter; iter = iter->next) {
+		m = iter->data;
+		bundle = m->component;
+		if (!list_search(candidates_for_removal, bundle, manifest_bundlename_strcmp)) {
```

Comment:
Instead of using list_search here can't we use `is_deletable_dependency()`?
And by that we can move this to a function in manifest.h `get_all_tracked_bundles()` (or a better name). This function can then be used to create a command to wipe out every bundle that isn't tracked (or dependency of a tracked bundle)

Reply:
No, that would not work. This loop is not just getting all the tracked bundles, it gets all installed bundles and their dependencies (which includes all tracked bundles and more), except for those we plan to remove. If we only get the tracked bundles and their dependencies it could uninstall bundles that are dependencies of other bundles that are going to remain installed, breaking them.

As an example consider the scenario from test REM026, in this scenario we have 7 bundles installed in the system:
 - test-bundle1 is tracked and includes os-core, test-bundle2, test-bundle4, test-bundle6, and test-bundle3 (via test-bundle2).
 - test-bundle5 is installed (but not tracked) and includes test-bundle4.

If we only had gotten the tracked bundles and their dependencies as you suggested and we run:
`$ swupd bundle-remove --recursive test-bundle1`
Then test-bundle4 will get removed, but it should not sine it is required by test-bundle5.

Reply:
This makes sense. We need to check all installed bundles that aren't going to be removed, not just the tracked ones. So this is correct.